src/Parsing/Impl/FStringParser.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.IO;
+using System.Text;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Diagnostics;
+using Microsoft.Python.Core.Text;
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Parsing {
+    public class FStringParser {
+        private readonly string _fString;
+        private readonly StringBuilder _buffer = new StringBuilder();
+        private readonly FStringBuilder _builder;
+        private readonly ErrorSink _errors;
+        private int _errorCode = 0;
+        private int _position = 0;
+
+        public FStringParser(FStringBuilder builder, string fString, ErrorSink errors) {
+            _fString = fString;
+            _builder = builder;
+            _errors = errors;
+        }
+
+        public void Parse() {
+            while (!EndOfFString()) {
+                if (IsDoubleBrace()) {
+                    NextChar();
+                    _buffer.Append(NextChar());
+                } else if (CurrentChar() == '{') {
+                    AddBufferedSubstring();
+                    ParseInnerExpression();
+                } else if (CurrentChar() == '}') {
+                    throw new Exception("closing '}' without opening");
+                } else {
+                    _buffer.Append(NextChar());
+                }
+            }
+            AddBufferedSubstring();
+        }
+
+        private bool IsDoubleBrace() {
+            StringSpan doubleOpen = new StringSpan("{{", 0, 2);
```

Comment:
Should be static/const to reduce amount of garbage in memory. No need to create objects for every comparo.